### ISSUE TITLE
realloc() performance: -d:useMalloc vs Nim default allocator

### ISSUE THREAD

**Issue by zevv** [nim-lang/Nim#13203]
Nim's default memory allocator does not do well on "incrementally growing" seqs or strings.

While going through the alloc code I noticed that Nim's default allocator does not try to grow buffers on `realloc()` but always does a fresh `alloc()`, copies the data and frees the old buffer. This has a quite some impact for some use cases, consider this snippet (Yes, I know I should use `repeat`)

```
var a: string
for i in 0..500000000:  a.add "01234567"
echo a.len
```

- Default alloc: `nim c --gc:arc -d:danger`: **13.0** seconds
- Libc malloc: `nim c --gc:arc -d:danger -d:useMalloc`: **4.0** seconds

**Comment by Araq** [comment]
Yeah, I know. The pre-ARC seq implementation could not use ``realloc`` so there was little incentive to make it fast.
